microcosm_pubsub/conventions.py

```diff
@@ -0,0 +1,98 @@
+"""
+Convention-driven pubsub.
+
+"""
+from enum import Enum, unique
+
+from inflection import underscore
+from marshmallow import fields
+
+from microcosm_pubsub.codecs import PubSubMessageSchema
+
+
+@unique
+class LifecycleChange(Enum):
+    """
+    CRUD lifecycle changes that may be announced via pubsub.
+
+    Note that `Changed` is strongly discouraged; use immutable resources instead.
+
+    """
+    Changed = u"changed"
+    Created = u"created"
+    Deleted = u"deleted"
+
+    def matches(self, media_type):
+        return self.value in media_type.split(".")
+
+
+def make_media_type(resource, lifecycle_change=None, organization=None, public=False):
+    """
+    Generate a canonical media type for a message that announces a lifecycle change to a resource.
+
+    :param resource: the resource (name)
+    :param lifecycle_change: the change that happened to the resource
+    :param organization: the vendor organization to which this resource belongs
+    :param public: whether the resource is meant for public/external consumption
+
+    """
+    if lifecycle_change is None:
+        lifecycle_change = LifecycleChange.Created
+    if organization is None:
+        organization = "globality"
+
+    return "application/vnd.{}.{}.{}.{}".format(
+        # use a vendor specific media type
+        organization,
+        # differentiate messages sent over a pubsub channel from other messages
+        "pubsub" if public else "pubsub._",
+        # qualify the message with the kind of lifecycle change
+        lifecycle_change.value,
+        # specify the resource name
+        underscore(resource),
+    )
+
+
+class URIMessageSchema(PubSubMessageSchema):
+    """
+    Define a baseline message schema that points to (the URI of) a resource that experienced a lifecycle change.
+
+    By convention, pubsub messages should be *references* to something that happened within some other
+    source-of-truth (e.g. a CRUD microservice). Because there are inherent race conditions between publishing
+    a message and commit a persistent transaction, message consumers are expected to call back to source-of-truth
+    (e.g. via HTTP) and fetch the current resource value. In the event that message has not been committed yet,
+    the consumer can retry a few times before giving up (e.g. via SQS dead-lettering).
+
+    """
+    def __init__(self, media_type, **kwargs):
+        super(PubSubMessageSchema, self).__init__(**kwargs)
+        self.MEDIA_TYPE = media_type
+
+    uri = fields.String(required=True)
```

Comment:
@jessemyers this is a pretty major use case - for the text highlighting for example right now we're intercepting all qna event pubsub messages and pre-filtering based on the 'tag' - otherwise we'd be materializing very many QnAEvents for no reason.. I think makes sense to have a notion of a 'header' for a pubsub message that can vary based on message type and contains information useful for pre-filtering - we can formalize that notion more carefully so is not as ad hoc as it was..